tests/unit/test_charm.py

```diff
@@ -612,17 +701,546 @@ def test_given_default_config_with_interfaces_when_network_attachment_definition
+    @patch("ops.model.Container.exec", new=MagicMock)
+    def test_given_upf_configured_to_run_in_dpdk_mode_when_create_network_attachment_definitions_then_2_nads_are_returned(  # noqa: E501
+        self, patch_get_service, kubernetes_create_object
+    ):
+        service_info_mock = Mock()
+        service_info_mock.is_running.return_value = True
+        patch_get_service.return_value = service_info_mock
+        self.harness.update_config(
+            key_values={
+                "upf-mode": "dpdk",
+                "enable-hugepages": True,
+                "access-interface-mac-address": VALID_ACCESS_MAC,
+                "core-interface-mac-address": VALID_CORE_MAC,
+            }
+        )
+
+        create_nad_calls = kubernetes_create_object.call_args_list
+        self.assertEqual(len(create_nad_calls), 2)
+
+    @patch("lightkube.core.client.Client.create")
+    @patch("ops.model.Container.get_service")
+    @patch("lightkube.core.client.GenericSyncClient", new=Mock)
+    @patch(
+        f"{HUGEPAGES_LIBRARY_PATH}.KubernetesHugePagesPatchCharmLib.is_patched",
+        Mock(return_value=True),
+    )
+    @patch(f"{MULTUS_LIBRARY_PATH}.KubernetesMultusCharmLib.is_ready", Mock(return_value=True))
+    @patch("charm.DPDK.is_configured", Mock(return_value=True))
+    @patch("ops.model.Container.exec", new=MagicMock)
+    def test_given_upf_configured_to_run_in_dpdk_mode_when_create_network_attachment_definitions_then_nad_type_is_vfioveth(  # noqa: E501
+        self, patch_get_service, kubernetes_create_object
+    ):
+        service_info_mock = Mock()
+        service_info_mock.is_running.return_value = True
+        patch_get_service.return_value = service_info_mock
+        self.harness.update_config(
+            key_values={
+                "upf-mode": "dpdk",
+                "enable-hugepages": True,
+                "access-interface-mac-address": VALID_ACCESS_MAC,
+                "core-interface-mac-address": VALID_CORE_MAC,
+            }
+        )
+
+        create_nad_calls = kubernetes_create_object.call_args_list
+        for create_nad_call in create_nad_calls:
+            create_nad_call_args = next(
+                iter(filter(lambda call_item: isinstance(call_item, dict), create_nad_call))
+            )
+            nad_config = json.loads(create_nad_call_args.get("obj").spec.get("config"))
+            self.assertEqual(nad_config["type"], "vfioveth")
+
+    @patch("lightkube.core.client.Client.create")
+    @patch("ops.model.Container.get_service")
+    @patch("lightkube.core.client.GenericSyncClient", new=Mock)
+    @patch(
+        f"{HUGEPAGES_LIBRARY_PATH}.KubernetesHugePagesPatchCharmLib.is_patched",
+        Mock(return_value=True),
+    )
+    @patch(f"{MULTUS_LIBRARY_PATH}.KubernetesMultusCharmLib.is_ready", Mock(return_value=True))
+    @patch("charm.DPDK.is_configured", Mock(return_value=True))
+    @patch("ops.model.Container.exec", new=MagicMock)
```

Comment:
We likely don't need to mock `exec` calls anymore as of PR #59

Reply:
True. Rebased and cleaned up.